Add resolution to Amcrest camera unique id

flacjacket opened the pull request:
## Proposed change

The current unique id assigned to the Amcrest camera is based on the device serial number. However, the camera supports multiple streams which have different resolutions. Add this information to the unique id to allow each stream to be configured as a separate camera.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #56126
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

flacjacket committed:
Add resolution to Amcrest camera unique id

janiversen commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -387,7 +387,9 @@ def update(self) -> None:
                     self._model = "unknown"
             if self._attr_unique_id is None:
-                self._attr_unique_id = self._api.serial_number.strip()
+                self._attr_unique_id = (
```
How will this work with existing configurations, I think you need to add some code that updates existing self._attr_unique_id.

flacjacket replied:
How would that be done? I'm not sure what changes are needed to update an existing unique id.

brianegge replied:
I don't think it can make it upgrade hass automatically. But you could mark the change as breaking.

janiversen replied:
It is possible, delete the old config entry and automatically make a new one.

You cannot update the unique id, but you can save people from having to manually delete and create a new configuration entry.

flacjacket replied:
That is great, do you know where that would need to be done or if there is an example of this somewhere else? Is that something that would need to be handled in `async_setup_platform`?

flacjacket committed:
Only use resolution in the unique id for low res stream

This makes this less of a breaking change for the default resolution as
the unique id of the camera will not change.

flacjacket commented:
I tweaked this a bit so it will only change the unique id if the camera is explicitly set to low resolution. This should hopefully limit the blast radius of the breaking change.

janiversen commented:
“ This may leave orphaned camera entities with the old unique id that would need to be manually deleted” is a sentence I really try to avoid.

You know in the code when there is a config entry with an old unique so at least delete it, so the user do not have to search (and once deleted it is pretty easy to recreate with the new id).

janiversen commented:
You should do it in async_setup_platform, there might be other places, but this one is secure.

flacjacket committed:
Update the entity on upgrade

flacjacket commented:
I added some migration logic that will look for the entity with the old unique id and if it exists then update it with the new unique id. Let me know if you think this is what you were thinking. I tested this by launching an HA instance with the camera configured on the current `dev` branch (which creates the entity with the previous unique id), then change to this branch and launch HA again. I see the update message fire and the camera entity persists through the update.

flacjacket commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -132,7 +135,22 @@ async def async_setup_platform(
-    async_add_entities([AmcrestCam(name, device, hass.data[DATA_FFMPEG])], True)
+    entity = AmcrestCam(name, device, hass.data[DATA_FFMPEG])
+
+    # 2021.9.0 introduced unique id's for the camera entity, but these were not
+    # unique for different resolution streams.  If any cameras were configured
+    # with this version, update the old entity with the new unique id.
```
I'm not sure what is the policy for deprecating this migration logic.

janiversen replied:
The only policy is to let it stay for a long time, because many do not upgrade every month. I just removed similar code in tradfri after 2 years.  In other words leave it until you have a good reason to remove it.

flacjacket commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -132,7 +135,22 @@ async def async_setup_platform(
+    # 2021.9.0 introduced unique id's for the camera entity, but these were not
+    # unique for different resolution streams.  If any cameras were configured
+    # with this version, update the old entity with the new unique id.
+    serial_number = await hass.async_add_executor_job(lambda: device.api.serial_number)  # type: ignore[no-any-return]
```
I think the `type: ignore` is needed here because of the lambda. Any thoughts on how to get mypy to know what this is would be appreciated.

janiversen replied:
I still have not seen a good solution, and according to google this is a major deficit.

ludeeus replied:
The value is striped everywhere else, why is that not needed here?

flacjacket replied:
Good catch, thanks. Added the `.strip()` for the serial number.

janiversen commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -132,7 +135,22 @@ async def async_setup_platform(
+    registry = entity_registry.async_get(hass)
+    old_entity_id = registry.async_get_entity_id(CAMERA_DOMAIN, DOMAIN, serial_number)
+    if old_entity_id is not None:
+        _LOGGER.info(f"Updating unique id for camera {old_entity_id}")
```
```suggestion
        msg  = f"Updating unique id for camera {old_entity_id}"
        _LOGGER.INFO(msg)
```
please do not use f”” in function calls.

flacjacket committed:
Address pylint problems

janiversen commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -132,7 +135,22 @@ async def async_setup_platform(
+    old_entity_id = registry.async_get_entity_id(CAMERA_DOMAIN, DOMAIN, serial_number)
+    if old_entity_id is not None:
+        _LOGGER.info(f"Updating unique id for camera {old_entity_id}")
+        await hass.async_add_executor_job(entity._update_unique_id)
```
Why is this needed? you change the entity a bit later

janiversen reviewed (changes requested):
Looks good.
1 suggestion, and 1 question.

janiversen commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -144,8 +144,8 @@ async def async_setup_platform(
     if old_entity_id is not None:
-        _LOGGER.info(f"Updating unique id for camera {old_entity_id}")
-        await hass.async_add_executor_job(entity._update_unique_id)
+        _LOGGER.info("Updating unique id for camera %s", old_entity_id)
```
please use f”” as I suggested.
The string used in a function call should be fix not variable.

janiversen replied:
```suggestion
        msg = f"Updating unique id for camera {old_entity_id}”
        _LOGGER.info(msg)
```

flacjacket replied:
Thanks for the suggested fix. I was able to roll this into the commit that fixes all of the pylint issues.

janiversen commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -144,8 +144,8 @@ async def async_setup_platform(
-        _LOGGER.info(f"Updating unique id for camera {old_entity_id}")
-        await hass.async_add_executor_job(entity._update_unique_id)
+        _LOGGER.info("Updating unique id for camera %s", old_entity_id)
+        await hass.async_add_executor_job(entity.update_unique_id)
```
this solves my question.

flacjacket replied:
Sure. I was thinking the unique id could be just computed in this function. It would duplicate where it is set between here and the `update` method, but it would simplify the `update` method to not have to call this function. If the unique id were something that is changing, then I feel like having the common helper is useful, but given that it shouldn't change and any change that we have will require some large migration like this, just duplicating the logic seems reasonable. What do you think about that?

janiversen replied:
I agree with you.

janiversen replied:
and btw I like your migration.

flacjacket replied:
Great, just pushed a new commit with the change here.

janiversen reviewed (changes requested):
please look at my suggestion in the previous reviewer

flacjacket committed:
Address comment on logging

flacjacket committed:
Remove call to update_unique_id

janiversen reviewed (approved):
LGTM.

Please remember to hit "resolve conversation", when you have pushed a commit that fixes a request for change. I (and I believe many others) check if there are open conversations before starting a review.

janiversen commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -387,8 +404,10 @@ def update(self) -> None:
+                serial_number = self._api.serial_number.strip()
+                if serial_number:
+                    self._attr_unique_id = f"{serial_number}-{self._resolution}"
+                    _LOGGER.debug("Assigned unique_id=%s", self._attr_unique_id)
```
```suggestion
                    msg = f"Assigned unique_id={self._attr_unique_id}¨
                    _LOGGER.debug(msg)
```

ludeeus replied:
For the Logger, we should use lazy formating and not assign a local constants with templates if we can

ludeeus commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -132,7 +135,21 @@ async def async_setup_platform(
+    registry = entity_registry.async_get(hass)
+    entity_id = registry.async_get_entity_id(CAMERA_DOMAIN, DOMAIN, serial_number)
+    if entity_id is not None:
+        msg = f"Updating unique id for camera {entity_id}"
+        _LOGGER.info(msg)
```
```suggestion
        _LOGGER.debug("Updating unique id for camera %s", entity_id)
```

janiversen replied:
Now this is interesting, I have received several review comments, telling me to always use f"" :smile:

ludeeus replied:
Generally yes, but not for the Logger

janiversen reviewed (changes requested):
sorry found one place.

flacjacket committed:
Change logging string formatting

Co-authored-by: Joakim Sørensen <hi@ludeeus.dev>

janiversen reviewed (approved):
LGTM

flacjacket committed:
Add channel now so another migration isn't needed later

flacjacket committed:
Update sensor and binary sensor use of channel to match camera

flacjacket commented:
@janiversen @ludeeus Anything else that is needed here? It would be great to get this bugfix in before 2021.10 betas start to get cut!

janiversen commented:
for me its OK, but I would like to hear the opinion og @ludeeus

flacjacket committed:
Strip serial number when resolving entity id

frenck commented:
Rebased the PR onto the latest `dev` to deal with the CI issue.

frenck reviewed (approved):
Thanks, @flacjacket 👍
